include/boost/url/query_param.hpp

```diff
@@ -136,17 +171,29 @@ struct query_param
 
 #ifndef BOOST_URL_DOCS
 #ifdef BOOST_NO_CXX14_AGGREGATE_NSDMI
+    template <class Allocator = std::allocator<char>>
     query_param(
-        string_view key_ = {},
-        string_view value_ = {},
-        bool has_value_ = false )
+        const_string key_,
+        const_string value_ = {},
+        bool has_value_ = false)
         : key(key_)
         , value(value_)
         , has_value(has_value_)
     {
     }
 #endif
 #endif
+
+private:
+    friend class params_view;
+    friend class params;
+
+    BOOST_URL_DECL
+    query_param(
```

Comment:
yeah shouldn't we just make this public?

Reply:
This overload

```
query_param(
        char const* s,
        std::size_t nk,
        std::size_t nv,
        const_string::factory const& a);
```

has no meaning to users. It's used by the params constructors to create query_param from a `raw_param`, which is an implementation detail.

In the future, we could have some overload like:

```
query_param(
        string_view k,
        string_view v,
        bool has_value,
        const_string::factory const& a);
```

but there's no demand for that in the library or from users yet.